### PR TITLE
Add cudnn.fla: a cuDNN drop-in for flash-linear-attention (GDN + KDA)

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
## What this is

`cudnn.fla` — a **drop-in accelerator for [flash-linear-attention](https://github.com/fla-org/flash-linear-attention) (FLA)**. One call monkeypatches the FLA ops cuDNN can serve to cuDNN's Blackwell (SM100) kernels, with a **transparent fallback** to FLA wherever cuDNN has no kernel, so results never change:

```python
import cudnn.fla
cudnn.fla.accelerate_fla()   # before importing FLA layers/models
import fla  # GatedDeltaNet / KDA now run on cuDNN where supported
```

Aligns with the `cudnn.torch` / `cudnn.jax` framework-integration folders.

## Ops accelerated today

- **Gated DeltaNet** (`chunk_gated_delta_rule`) — the GDN convention (g log-space decay, beta post-sigmoid, GVA where `HV > H`) mapped onto cuDNN's native op; the fused-layer knobs (`use_gate_in_kernel`, `use_beta_sigmoid_in_kernel`, `use_qk_l2norm_in_kernel`) are reproduced, using FLA's own `l2norm` kernel (torch `F.normalize` is ~2.6x slower).
- **KDA** (Kimi Delta Attention, `chunk_kda`) — channel-wise gate + scalar beta; l2norm fwd+bwd through cuDNN, beta-sigmoid/safe-gate in the shim. **bf16-only** (fp16 -> NaN -> the shim declines and falls back).

Configs cuDNN cannot serve raise `cudnnGraphNotSupportedError` / `NotImplementedError` and fall back to FLA — never a wrong answer.

## Correctness

`test/python/linear_attention/test_fla_compat.py` (L0, SM100-gated) requires cuDNN to match FLA within FLA's **own** bf16 noise on the output and every gradient (fp32-naive reference, ratio test); a config that doesn't match must fall back rather than run. All pass on B200. Includes GDN dense/h8/fp16/GVA, the fused-layer path, KDA (bf16), and fallback/patch-restore tests.

## Also

- `benchmark/linear_attention/fla_e2e_perf_share.py` — an end-to-end hybrid Gated DeltaNet LM perf-share / support-gap benchmark.

## Not yet

`gated_delta_net_v2` (GDN-2) shim; varlen/THD parity; `import cudnn.fla` install-time mirror. Tracked separately.

---
<sub>note to self: claude::11323ca1-07bc-4fc4-8ec7-ba95d8f061d8 — cudnn.fla shim. cwd /home/scratch.yanxu_libs/cudnn_frontend</sub>

## Summary by CodeRabbit

* **New Features**
  * Added optional cuDNN acceleration for FLA gated delta rule and KDA operations.
  * Automatically falls back to native FLA implementations when configurations or hardware are unsupported.
  * Added status reporting and controls to enable, disable, and restore acceleration.
  * Added support for variable-length sequences, recurrent states, normalization, and fused gating options.

* **Tests**
  * Added compatibility and parity tests covering accelerated, fallback, gradient, and lifecycle behavior.

### CODE REVIEW

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost

**Comment by YangXu1990uiuc** [comment]
Addressed in the latest commit: KDA reshapes `A_log`/`dt_bias` with `H` (not `HO=max(H,HV)`) to match `g`'s `[B,T,H,K]` layout, with element-count validation that declines (falls back) instead of crashing on a mismatched GVA layout; on `safe_gate`, decline when `lower_bound` is omitted rather than guessing -5.0; `restore_fla` sets the owning module attribute back explicitly; benchmark rejects `attn_every < 1` and labels the host gap approximate; and the non-deterministic KDA gate-parameter gradients (dg/dA_log/dt_bias, cross-CTA atomicAdd) get a wider test slack than the data grads (removes a ~1/4 flake). Verified on B200.

Skipped, with reasons: (1) *backward-decline handling in the GDN path* — the shim commits forward+backward atomically through the native `gated_delta_net` autograd op; a forward that runs native has consistent native backward support (the fwd+bwd parity tests confirm it), and intercepting a decline inside `autograd.backward` to fall back isn't feasible without also having run FLA's forward. (2) *`cudnn.fla` exposure nit* — minor, no concrete change.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run frost

**Comment by YangXu1990uiuc** [comment]
Rebased onto `develop` @ #616 and switched the shims to #616's in-kernel fusion.

#616 ("fully support all L2norm, beta, gate fusion") added in-kernel L2-norm / beta-sigmoid / safe-gate to `gated_delta_net` and `kimi_delta_attention` (fwd+bwd). The GDN/KDA shims previously reproduced FLA's `use_*_in_kernel` transforms in torch and called the op with fusion off; they now forward the flags (`use_qk_l2norm_in_kernel`, `use_beta_sigmoid_in_kernel`, `safe_gate` + `a_log`/`dt_bias`). `safe_gate` computes `-exp(a_log)*softplus(g+dt_bias)`, matching FLA exactly.

Parity (`test_fla_compat.py`) is green on the output and every gradient, on the plain and fused-layer paths, bf16 + fp16.

Full-fat B200, CUDA-graph kernel time, the FLA `GatedDeltaNet` layer's fused call:

| shape | before (torch un-fuse) | after (in-kernel) |
|---|---|---|
| T2048 H16 | 1.94x (unstable across processes) | **2.34x (stable)** |
| T4096 H16 | 2.76x | **2.87x** |
| bs4 T2048 | 2.43x | **2.47x** |

Full `GatedDeltaNet` layer (hidden=2048): the earlier 0.77x is resolved → 1.00x (still projection-bound at this size; 1.27x eager from fewer launches).

KDA note: `kimi_delta_attention` fuses `safe_gate` but not the non-safe `-exp*softplus` channel gate that FLA's KDA layer drives via `use_gate_in_kernel`; the shim keeps that gate in torch, so KDA wins at small T (1.65x at T512) but loses at T>=2k (0.87x). Flagged as a kernel follow-up.

<sub>note to self: claude::11323ca1-07bc-4fc4-8ec7-ba95d8f061d8 — cwd /home/scratch.yanxu_libs/cudnn_frontend · worktree /home/scratch.yanxu_gpu/fe-fla-compat</sub>

**Review by jhjpark — APPROVED** [review]

File: python/cudnn/fla/__init__.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+requires cuDNN to match FLA within its own bf16 noise on the output and every
+gradient; a config that does not match must fall back rather than run.
+"""
+
+from __future__ import annotations
+
+import sys
+
+from .gated_delta_rule import make_chunk_gated_delta_rule, last_path
+from .kda import make_chunk_kda
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply python/cudnn/fla/__init__.py:25]: Applied the lazy `fla` export in `python/cudnn/__init__.py` (commit a4211b9), mirroring the existing `jax` / `experimental` branches, so `import cudnn; cudnn.fla.accelerate_fla()` resolves without a separate `import cudnn.fla`. It stays behind `__getattr__`, so `import cudnn` still never imports torch or the FLA shim (the import fires only on attribute access).

Deferring the docs page. `docs/fe-oss-apis/` is specifically the **CuTeDSL-kernel** FE OSS APIs — every page follows the wrapper + class-API template and the `[cutedsl]` optional dependency (the folder never references `linear_attention`/`gated_delta`/`kda`). `cudnn.fla` isn't a CuTeDSL kernel: it's a monkeypatch shim over the native `cudnn.linear_attention` ops and imports only torch, so it has neither a wrapper/class API nor a `[cutedsl]` dependency and wouldn't fit that template. Usage is covered by the module docstring; a dedicated OSS-API page for the shim is out of scope for this PR.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run python_tests
